cpp/benchmarks/stochastic_algorithm.cc

```diff
@@ -0,0 +1,131 @@
+#include <sopt/power_method.h>
+#include <sopt/relative_variation.h>
+#include <sopt/utilities.h>
+#include <sopt/wavelets.h>
+#include <sopt/wavelets/sara.h>
+
+#ifdef PURIFY_H5
+#include "purify/h5reader.h"
+#endif
+
+using namespace purify;
+
+class StochasticAlgoFixture : public ::benchmark::Fixture {
+ public:
+  void SetUp(const ::benchmark::State &state) {
+    m_imsizex = state.range(0);
+    m_imsizey = state.range(0);
+
+    m_sigma = 0.016820222945913496 * std::sqrt(2);  // see test_parameters file
+    m_beta = m_sigma * m_sigma;
+    m_gamma = 0.0001;
+
+    m_N = state.range(1);
+
+    m_input_data_path = data_filename("expected/fb/input_data.h5");
+
+    m_world = sopt::mpi::Communicator::World();
+  }
+
+  void TearDown(const ::benchmark::State &state) {}
+
+  sopt::mpi::Communicator m_world;
+
+  std::string m_input_data_path;
+
+  t_uint m_imsizey;
+  t_uint m_imsizex;
+
+  t_real m_sigma;
+  t_real m_beta;
+  t_real m_gamma;
+
+  size_t m_N;
+
+  std::vector<std::tuple<std::string, t_uint>> const m_sara{
+      std::make_tuple("Dirac", 3u), std::make_tuple("DB1", 3u), std::make_tuple("DB2", 3u),
+      std::make_tuple("DB3", 3u),   std::make_tuple("DB4", 3u), std::make_tuple("DB5", 3u),
+      std::make_tuple("DB6", 3u),   std::make_tuple("DB7", 3u), std::make_tuple("DB8", 3u)};
+};
+
+BENCHMARK_DEFINE_F(StochasticAlgoFixture, ForwardBackward)(benchmark::State &state) {
+  // This functor would be defined in Purify
+  std::function<std::shared_ptr<sopt::IterationState<Vector<t_complex>>>()> random_updater =
+      [this]() {
+        H5::H5Handler h5file(m_input_data_path, m_world);
+        utilities::vis_params uv_data = H5::stochread_visibility(h5file, m_N, false);
+        uv_data.units = utilities::vis_units::radians;
+        auto phi = factory::measurement_operator_factory<Vector<t_complex>>(
+            factory::distributed_measurement_operator::mpi_distribute_image, uv_data, m_imsizex,
+            m_imsizey, 1, 1, 2, kernels::kernel_from_string.at("kb"), 4, 4);
+
+        return std::make_shared<sopt::IterationState<Vector<t_complex>>>(uv_data.vis, phi);
+      };
+
+  auto IS = random_updater();
+  auto Phi = IS->Phi();
+
+  auto const power_method_stuff = sopt::algorithm::power_method<Vector<t_complex>>(
+      Phi, 1000, 1e-5, m_world.broadcast(Vector<t_complex>::Ones(m_imsizex * m_imsizey).eval()));
+
+  const t_real op_norm = std::get<0>(power_method_stuff);
+
+  // wavelets
+  auto const wavelets = factory::wavelet_operator_factory<Vector<t_complex>>(
+      factory::distributed_wavelet_operator::serial, m_sara, m_imsizey, m_imsizex);
+
+  // algorithm
+  sopt::algorithm::ImagingForwardBackward<t_complex> fb(random_updater);
+  fb.itermax(state.range(2))
+      .step_size(m_beta * sqrt(2))
+      .sigma(m_sigma * sqrt(2))
+      .regulariser_strength(m_gamma)
+      .relative_variation(1e-3)
+      .residual_tolerance(0)
+      .tight_frame(true)
+      .sq_op_norm(op_norm * op_norm)
```

Comment:
I think the op norm needs to be moved to the linear operator now?

Reply:
Indeed. It helps when you build the benchmark